### PR TITLE
Added external spi flash driver.

### PR DESCRIPTION
**Pull request by wb-Joy**
Added external spi flash driver.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [x] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by wb-Joy** [comment]
@tzarc  Describe where the document should be added?

**Comment by drashna** [comment]
presumably: https://github.com/qmk/qmk_firmware/blob/develop/docs/eeprom_driver.md

**Comment by tzarc** [comment]
Almost, will need a `flash_driver.md` with descriptions of configurable items, as well as a link to the new page in the sidebar.

**Comment by drashna** [comment]
> Almost, will need a `flash_driver.md` with descriptions of configurable items, as well as a link to the new page in the sidebar.

I should read more.   But should be set up somewhat similarly, I think.  At least in the sidebar for the docs, and such.

File: docs/ja/_summary.md [hunk 1]
```diff
@@ -135,6 +135,7 @@
       * [SPI ドライバ](ja/spi_driver.md)
       * [WS2812 ドライバ](ja/ws2812_driver.md)
       * [EEPROM ドライバ](ja/eeprom_driver.md)
+      * [FLASH ドライバ](ja/flash_driver.md)
```
Comment (tzarc) [inline docs/ja/_summary.md:138]: Please rollback this change and leave this update to the translators.

File: docs/zh-cn/_summary.md [hunk 2]
```diff
@@ -107,6 +107,7 @@
   * [SPI设备](zh-cn/spi_driver.md)
   * [WS2812设备](zh-cn/ws2812_driver.md)
   * [EEPROM设备](zh-cn/eeprom_driver.md)
+  * [FLASH设备](zh-cn/flash_driver.md)
```
Comment (tzarc) [inline docs/zh-cn/_summary.md:110]: Please rollback this change and leave this update to the translators.

File: drivers/flash/flash_spi.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    the intended divisor -- check your clock settings and the datasheet of
+    your FLASH.
+*/
+#ifndef EXTERNAL_FLASH_SPI_CLOCK_DIVISOR
+#    ifdef __AVR__
+#        define EXTERNAL_FLASH_SPI_CLOCK_DIVISOR 8
+#    else
+#        define EXTERNAL_FLASH_SPI_CLOCK_DIVISOR 18
+#    endif
+#endif
```
Comment (tzarc) [inline drivers/flash/flash_spi.h:42]:
```suggestion
#ifndef EXTERNAL_FLASH_SPI_CLOCK_DIVISOR
#    ifdef __AVR__
#        define EXTERNAL_FLASH_SPI_CLOCK_DIVISOR 4
#    else
#        define EXTERNAL_FLASH_SPI_CLOCK_DIVISOR 8
#    endif
#endif
```
This should be a power of 2. The reference MCU used is ATMEGA32U4 for AVRs, and STM32F303 for ARMs. The above values will result in 4MHz clock for 32U4, and 4.5MHz clock for F303 with the default board definitions.

Keyboards are then free to override as they see fit.

File: drivers/flash/flash_spi.c [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+
+    spi_write(FLASH_CMD_WRDI);
+
+    spi_stop();
+
+    return FLASH_STATUS_SUCCESS;
+}
+
+/* This function is used for read transfer, write transfer and erase transfer. */
+static flash_status_t spi_flash_transmit(uint8_t cmd, uintptr_t addr, uint8_t *data, size_t len) {
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:128]:
```suggestion
static flash_status_t spi_flash_transaction(uint8_t cmd, uint32_t addr, uint8_t *data, size_t len) {
```

File: drivers/flash/flash_spi.c [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    response= spi_flash_wait_while_busy();
+    if (response != FLASH_STATUS_SUCCESS) {
+        dprint("Failed to check WIP flag! [spi flash erase chip]\n");
+        return response;
+    }
+
+    return response;
+}
+
+flash_status_t flash_erase_sector(const void *addr) {
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:205]:
```suggestion
flash_status_t flash_erase_sector(uint32_t addr) {
```
Please also swap other usages of `target_addr` to just `addr`.

File: drivers/flash/flash_spi.c [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+    flash_status_t response    = FLASH_STATUS_SUCCESS;
+    uintptr_t      target_addr = (uintptr_t)addr;
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:207]:
```suggestion
```

File: drivers/flash/flash_spi.c [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+    response= spi_flash_wait_while_busy();
+    if (response != FLASH_STATUS_SUCCESS) {
+        dprint("Failed to check WIP flag! [spi flash erase sector]\n");
+        return response;
+    }
+
+    return response;
+}
+
+flash_status_t flash_erase_block(const void *addr) {
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:247]:
```suggestion
flash_status_t flash_erase_block(uint32_t addr) {
```
As per previous, swap over usages of `target_addr`.

File: drivers/flash/flash_spi.c [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+    flash_status_t response    = FLASH_STATUS_SUCCESS;
+    uintptr_t      target_addr = (uintptr_t)addr;
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:249]:
```suggestion
```

File: drivers/flash/flash_spi.c [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+    response= spi_flash_wait_while_busy();
+    if (response != FLASH_STATUS_SUCCESS) {
+        dprint("Failed to check WIP flag! [spi flash erase block]\n");
+        return response;
+    }
+
+    return response;
+}
+
+flash_status_t flash_read_block(uint8_t *buf, const void *addr, size_t len) {
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:289]:
```suggestion
flash_status_t flash_read_block(uint32_t addr, void *buf, size_t len) {
```
As per previous, swap over usages of `target_addr`.

File: drivers/flash/flash_spi.c [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+    flash_status_t response    = FLASH_STATUS_SUCCESS;
+    uint8_t *      read_buf    = (uint8_t *)buf;
+    uintptr_t      target_addr = (uintptr_t)addr;
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:292]:
```suggestion
```

File: drivers/flash/flash_spi.c [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+    for (size_t i = 0; i < len; ++i) {
+        dprintf(" %02X", (int)(((uint8_t *)read_buf)[i]));
+    }
+    dprintf("\n");
+#endif  // DEBUG_FLASH_SPI_OUTPUT
+
+    return response;
+}
+
+flash_status_t flash_write_block(const uint8_t *buf, void *addr, size_t len) {
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:321]:
```suggestion
flash_status_t flash_write_block(uint32_t addr, const void *buf, size_t len) {
```
As per previous, swap over usages of `target_addr`.

File: drivers/flash/flash_spi.c [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+    flash_status_t response    = FLASH_STATUS_SUCCESS;
+    uint8_t *      write_buf   = (uint8_t *)buf;
+    uintptr_t      target_addr = (uintptr_t)addr;
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:324]:
```suggestion
```

File: drivers/flash/flash_spi.h [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+#ifndef EXTERNAL_FLASH_PAGE_SIZE
+#    define EXTERNAL_FLASH_PAGE_SIZE 256
+#endif
+
+/*
+    The sector size of the FLASH in Kbytes, as specified in the datasheet.
+*/
+#ifndef EXTERNAL_FLASH_SECTOR_SIZE
+#    define EXTERNAL_FLASH_SECTOR_SIZE 4
+#endif
```
Comment (tzarc) [inline drivers/flash/flash_spi.h:78]:
```suggestion
/*
    The sector size of the FLASH in bytes, as specified in the datasheet.
*/
#ifndef EXTERNAL_FLASH_SECTOR_SIZE
#    define EXTERNAL_FLASH_SECTOR_SIZE (4 * 1024)
#endif
```

File: drivers/flash/flash_spi.h [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+
+/*
+    The block size of the FLASH in Kbytes, as specified in the datasheet.
+*/
+#ifndef EXTERNAL_FLASH_BLOCK_SIZE
+#    define EXTERNAL_FLASH_BLOCK_SIZE 64
+#endif
```
Comment (tzarc) [inline drivers/flash/flash_spi.h:85]:
```suggestion
/*
    The block size of the FLASH in bytes, as specified in the datasheet.
*/
#ifndef EXTERNAL_FLASH_BLOCK_SIZE
#    define EXTERNAL_FLASH_BLOCK_SIZE (64 * 1024)
#endif
```

File: drivers/flash/flash_spi.h [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+
+/*
+    The total size of the FLASH in Kbytes, as specified in the datasheet.
+*/
+#ifndef EXTERNAL_FLASH_SIZE
+#    define EXTERNAL_FLASH_SIZE 512
+#endif
```
Comment (tzarc) [inline drivers/flash/flash_spi.h:92]:
```suggestion
/*
    The total size of the FLASH in bytes, as specified in the datasheet.
*/
#ifndef EXTERNAL_FLASH_SIZE
#    define EXTERNAL_FLASH_SIZE (512 * 1024)
#endif
```

File: drivers/flash/flash_spi.h [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+
+/*
+    The sector count of the FLASH, calculated by total FLASH size and sector size.
+*/
+#define EXTERNAL_FLASH_SECTOR_COUNT (EXTERNAL_FLASH_SIZE / EXTERNAL_FLASH_SECTOR_SIZE)
+
+/*
+    The page count of the FLASH, calculated by total FLASH size and page size.
+*/
+#define EXTERNAL_FLASH_PAGE_COUNT ((EXTERNAL_FLASH_SIZE << 10) / EXTERNAL_FLASH_PAGE_SIZE)
```
Comment (tzarc) [inline drivers/flash/flash_spi.h:107]:
```suggestion
#define EXTERNAL_FLASH_PAGE_COUNT ((EXTERNAL_FLASH_SIZE) / (EXTERNAL_FLASH_PAGE_SIZE))
```

File: drivers/flash/flash_spi.h [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+
+/*
+    The block count of the FLASH, calculated by total FLASH size and block size.
+*/
+#define EXTERNAL_FLASH_BLOCK_COUNT (EXTERNAL_FLASH_SIZE / EXTERNAL_FLASH_BLOCK_SIZE)
+
+/*
+    The sector count of the FLASH, calculated by total FLASH size and sector size.
+*/
+#define EXTERNAL_FLASH_SECTOR_COUNT (EXTERNAL_FLASH_SIZE / EXTERNAL_FLASH_SECTOR_SIZE)
```
Comment (tzarc) [inline drivers/flash/flash_spi.h:102]:
```suggestion
#define EXTERNAL_FLASH_SECTOR_COUNT ((EXTERNAL_FLASH_SIZE) / (EXTERNAL_FLASH_SECTOR_SIZE))
```

File: drivers/flash/flash_spi.h [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+    The total size of the FLASH in Kbytes, as specified in the datasheet.
+*/
+#ifndef EXTERNAL_FLASH_SIZE
+#    define EXTERNAL_FLASH_SIZE 512
+#endif
+
+/*
+    The block count of the FLASH, calculated by total FLASH size and block size.
+*/
+#define EXTERNAL_FLASH_BLOCK_COUNT (EXTERNAL_FLASH_SIZE / EXTERNAL_FLASH_BLOCK_SIZE)
```
Comment (tzarc) [inline drivers/flash/flash_spi.h:97]:
```suggestion
#define EXTERNAL_FLASH_BLOCK_COUNT ((EXTERNAL_FLASH_SIZE) / (EXTERNAL_FLASH_BLOCK_SIZE))
```

File: drivers/flash/flash_spi.c [hunk 19]
```diff
@@ -0,0 +1,382 @@
...
+    return response;
+}
+
+flash_status_t flash_erase_sector(const void *addr) {
+    flash_status_t response    = FLASH_STATUS_SUCCESS;
+    uintptr_t      target_addr = (uintptr_t)addr;
+
+    /* Check that the address exceeds the limit. */
+    if ((target_addr + (EXTERNAL_FLASH_SECTOR_SIZE << 10)) >= (EXTERNAL_FLASH_SIZE << 10) ||
+        ((target_addr % (EXTERNAL_FLASH_SECTOR_SIZE << 10)) != 0)) {
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:211]:
```suggestion
    if ((target_addr + (EXTERNAL_FLASH_SECTOR_SIZE)) >= (EXTERNAL_FLASH_SIZE) ||
        ((target_addr % (EXTERNAL_FLASH_SECTOR_SIZE)) != 0)) {
```

File: drivers/flash/flash_spi.c [hunk 20]
```diff
@@ -0,0 +1,382 @@
...
+    return response;
+}
+
+flash_status_t flash_erase_block(const void *addr) {
+    flash_status_t response    = FLASH_STATUS_SUCCESS;
+    uintptr_t      target_addr = (uintptr_t)addr;
+
+    /* Check that the address exceeds the limit. */
+    if ((target_addr + (EXTERNAL_FLASH_BLOCK_SIZE << 10)) >= (EXTERNAL_FLASH_SIZE << 10) ||
+        ((target_addr % (EXTERNAL_FLASH_BLOCK_SIZE << 10)) != 0)) {
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:253]:
```suggestion
    if ((target_addr + (EXTERNAL_FLASH_BLOCK_SIZE)) >= (EXTERNAL_FLASH_SIZE) ||
        ((target_addr % (EXTERNAL_FLASH_BLOCK_SIZE)) != 0)) {
```

**Review by tzarc — CHANGES_REQUESTED** [review]
As far as I can tell, all the "Kbyte" values have been converted to "byte", please confirm with hardware.

**Comment by tzarc** [comment]
Please also confirm that usages of `uintptr_t` have been changed to `uint32_t`.

File: docs/flash_driver.md [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+
+`config.h` override                            | Description                                                                          | Default Value
+-----------------------------------------------|--------------------------------------------------------------------------------------|--------------
+`#define EXTERNAL_FLASH_SPI_SLAVE_SELECT_PIN`  | SPI Slave select pin in order to inform that the FLASH is currently being addressed  | _none_
+`#define EXTERNAL_FLASH_SPI_CLOCK_DIVISOR`     | Clock divisor used to divide the peripheral clock to derive the SPI frequency        | `18`
+`#define EXTERNAL_FLASH_PAGE_SIZE`             | The Page size of the FLASH in bytes, as specified in the datasheet                   | 256
+`#define EXTERNAL_FLASH_SECTOR_SIZE`           | The sector size of the FLASH in Kbytes, as specified in the datasheet                | 4
+`#define EXTERNAL_FLASH_BLOCK_SIZE`            | The block size of the FLASH in Kbytes, as specified in the datasheet                 | 64
+`#define EXTERNAL_FLASH_SIZE`                  | The total size of the FLASH in Kbytes, as specified in the datasheet                 | 512
```
Comment (tzarc) [inline docs/flash_driver.md:21]:
```suggestion
`#define EXTERNAL_FLASH_SECTOR_SIZE`           | The sector size of the FLASH in bytes, as specified in the datasheet                 | `(4 * 1024)`
`#define EXTERNAL_FLASH_BLOCK_SIZE`            | The block size of the FLASH in bytes, as specified in the datasheet                  | `(64 * 1024)`
`#define EXTERNAL_FLASH_SIZE`                  | The total size of the FLASH in bytes, as specified in the datasheet                  | `(512 * 1024)`
```

File: common_features.mk [hunk 22]
```diff
@@ -201,6 +201,21 @@ else
...
 endif
 
+VALID_FLASH_DRIVER_TYPES := spi
+FLASH_DRIVER ?= no
+ifneq ($(strip $(FLASH_DRIVER)), no)
+    ifeq ($(filter $(FLASH_DRIVER),$(VALID_FLASH_DRIVER_TYPES)),)
+        $(error FLASH_DRIVER="$(FLASH_DRIVER)" is not a valid FLASH driver)
+    else
+        OPT_DEFS += -DFLASH_ENABLE
+        ifeq ($(strip $(EEPROM_DRIVER)), spi)
```
Comment (tzarc) [inline common_features.mk:211]:
```suggestion
        ifeq ($(strip $(FLASH_DRIVER)), spi)
```

File: drivers/flash/flash_spi.c [hunk 23]
```diff
@@ -0,0 +1,372 @@
...
+
+    /* Enable writes. */
+    response = spi_flash_write_enable();
+    if (response != FLASH_STATUS_SUCCESS) {
+        dprint("Failed to write-enable! [spi flash erase sector]\n");
+        return response;
+    }
+
+    /* Erase Sector. */
+    response = spi_flash_transmit((uint8_t)FLASH_CMD_SE, addr, NULL, 0);
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:225]:
```suggestion
    response = spi_flash_transaction(FLASH_CMD_SE, addr, NULL, 0);
```

File: drivers/flash/flash_spi.c [hunk 24]
```diff
@@ -0,0 +1,372 @@
...
+
+    /* Enable writes. */
+    response = spi_flash_write_enable();
+    if (response != FLASH_STATUS_SUCCESS) {
+        dprint("Failed to write-enable! [spi flash erase block]\n");
+        return response;
+    }
+
+    /* Erase Block. */
+    response = spi_flash_transmit((uint8_t)FLASH_CMD_BE, addr, NULL, 0);
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:265]:
```suggestion
    response = spi_flash_transaction(FLASH_CMD_BE, addr, NULL, 0);
```

File: drivers/flash/flash_spi.c [hunk 25]
```diff
@@ -0,0 +1,372 @@
...
+    /* Wait for the write-in-progress bit to be cleared. */
+    response = spi_flash_wait_while_busy();
+    if (response != FLASH_STATUS_SUCCESS) {
+        dprint("Failed to check WIP flag! [spi flash read block]\n");
+        memset(read_buf, 0, len);
+        return response;
+    }
+
+    /* Perform read. */
+    response = spi_flash_transmit((uint8_t)FLASH_CMD_READ, (uint32_t)addr, read_buf, len);
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:294]:
```suggestion
    response = spi_flash_transaction(FLASH_CMD_READ, addr, read_buf, len);
```

File: drivers/flash/flash_spi.c [hunk 26]
```diff
@@ -0,0 +1,372 @@
...
+    if (response != FLASH_STATUS_SUCCESS) {
+        dprint("Failed to read block! [spi flash read block]\n");
+        memset(read_buf, 0, len);
+        return response;
+    }
+
+#if defined(CONSOLE_ENABLE) && defined(DEBUG_FLASH_SPI_OUTPUT)
+    dprintf("[SPI FLASH R] 0x%08lX: ", ((uint32_t)addr));
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:302]:
```suggestion
    dprintf("[SPI FLASH R] 0x%08lX: ", addr);
```

File: drivers/flash/flash_spi.c [hunk 27]
```diff
@@ -0,0 +1,372 @@
...
+
+        /* Enable writes. */
+        response = spi_flash_write_enable();
+        if (response != FLASH_STATUS_SUCCESS) {
+            dprint("Failed to write-enable! [spi flash write block]\n");
+            return response;
+        }
+
+#if defined(CONSOLE_ENABLE) && defined(DEBUG_FLASH_SPI_OUTPUT)
+        dprintf("[SPI FLASH W] 0x%08lX: ", ((uint32_t)addr));
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:338]:
```suggestion
        dprintf("[SPI FLASH W] 0x%08lX: ", addr);
```

File: drivers/flash/flash_spi.c [hunk 28]
```diff
@@ -0,0 +1,372 @@
...
+        for (size_t i = 0; i < write_length; i++) {
+            dprintf(" %02X", (int)(uint8_t)(write_buf[i]));
+        }
+        dprintf("\n");
+#endif  // DEBUG_FLASH_SPI_OUTPUT
+
+        /* Perform the write. */
+        response = spi_flash_transmit((uint8_t)FLASH_CMD_PP, addr, write_buf, write_length);
```
Comment (tzarc) [inline drivers/flash/flash_spi.c:346]:
```suggestion
        response = spi_flash_transaction(FLASH_CMD_PP, addr, write_buf, write_length);
```

**Review by tzarc — CHANGES_REQUESTED** [review]

**Review by tzarc — APPROVED** [review]

**Review by zvecr — APPROVED** [review]

**Pull request merged**